apps/docs/refactor.md

```diff
@@ -0,0 +1,565 @@
+assumes that at some point the authentication library has been configured
+properly by the entry point. This would be unintentional coupling.
+
+Instead, the entry point should be responsible for passing the configured
+authentication selectors directly to the stepper library via it's `configure`
+function. Here is how that might look:
+
+**`src/entry-points/some-page.js`**
+Now our entry point has to configure both the `stepper` and the `auth`
+libraries and manage how they access each other's shared state.
+
+```js
+import {createStore, combineReducers} from 'redux';
+import {* as auth} from '../auth';
+import {* as stepper} from '../stepper';
+
+const store = createStore(
+  combineReducers({
+    authState: auth.reducer,
+    stepperState: stepper.reducer
+  })
+);
+
+auth.configure({
+  getState: state => state.authState,
+});
+stepper.configure({
+  getState: state => state.stepperState,
+  authSelectors: auth.selectors,
+});
+```
+
+**`src/stepper/index.js`**
+
+Meanwhile, the configure function of our `stepper` library would now accept a
+set of auth selectors. It doesn't need to know where these selectors came from,
+which has a nice side effect of making the `stepper` library easier to test with
+a "fake" set of auth selectors.
+
+```js
+import reducer from './redux/reducer';
+import {INCREMENT_STEP} from './redux/actionTypes';
+import {incrementStep, setAuthSelector} from './redux/actionCreators';
+import {getStep, setRootSelector} from './redux/selectors';
+
+export reducer;
+export const actionTypes = {INCREMENT_STEP}
+export const actions = {incrementStep};
+export const selectors = {getStep};
+
+export function configure(config) {
+  setRootSelector(config.getState);
+  setAuthSelector(config.authSelectors);
+}
+```
+
+**`src/stepper/redux/actionCreators.js`**
+
+Now we can update our action creators to check the login state before actually
+dispatching the action.
+
+```js
+import * as actionTypes from './actionTypes';
+
+let authSelectors = {
+  isLoggedIn: () => throw new Error("Auth selectors not configured.");
+};
+
+export function setAuthSelectors(newAuthSelectors) {
+  authSelectors = newAuthSelectors;
+}
+
+export function incrementStep() {
+  return (dispatch, getState) => {
+    if (authSelectors.isLoggedIn(getState())) {
+      dispatch({type: actionTypes.INCREMENT_STEP});
+    } else {
+      throw new Error("You can't increment a step without being logged in!");
+    }
+  }
+}
+
+export function decrementStep() {
+  return {type: actionTypes.DECREMENT_STEP}
+}
+```
+
+### Feature Library Rules ###
+
+In order to maintain some sanity, there are a few important rules that govern
+what feature libraries cannot do:
+
+- do not create redux stores
+- do not reference DOM nodes by ID or css selectors except for those DOM nodes
+  that were created and rendered to the DOM by the library.
+- do not depend directly on other feature libraries unless you are reponsible
+  for configuring them
```

Comment:
`s/you are/it was`?  This might need to be rephrased.  I think feature libraries _do_ depend on other feature libraries' interfaces but those interfaces _must_ be injected, not imported (which makes them interchangeable).  I'd also prefer that no feature library configures another feature library - only entry points do that.  Is there a particular use case for 'nested' feature libraries you're considering?